geomfn: consolidate affine transformations

otan opened the pull request:
Resolves #53646

* Introduce `applyOnGeomT`, which applies a function on a set
  of coordinates.
* Consolidate Translate, Scale and Rotate in terms of Affine.
  This is what PostGIS does, and removes a lot of duplicate code.
* Implement ST_Affine.
* Add unimplemented errors for remaining ST_Rotate items.

Release justification: low risk updates to to new functionality.

Release note (sql change): Implement ST_Affine.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53856)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53856#-:-MGKA2E37SIyignf7w26:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[docs/generated/sql/functions.md, line 1239 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53856#-MGK5q6eADGsYFgSeVpf:-MGK5q6eADGsYFgSeVpg:bszmzn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c55485bc3ce493c80bf989d855bbf95b581aef3/docs/generated/sql/functions.md#L1239)):*
> ```Markdown
> </span></td></tr>
> <tr><td><a name="st_affine"></a><code>st_affine(geometry: geometry, a: <a href="float.html">float</a>, b: <a href="float.html">float</a>, d: <a href="float.html">float</a>, e: <a href="float.html">float</a>, x_off: <a href="float.html">float</a>, y_off: <a href="float.html">float</a>) &rarr; geometry</code></td><td><span class="funcdesc"><p>Applies a 2D affine transformation to the given geometry.</p>
> <p>The matrix transformation will be apply as follows for each coordinate:
> ```

s/apply/applied/
___
*[pkg/geo/geomfn/affine_transforms.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53856#-MGK6X3h7Rp8hBZGDFn-:-MGK6X3h7Rp8hBZGDFn0:bwwjvgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c55485bc3ce493c80bf989d855bbf95b581aef3/pkg/geo/geomfn/affine_transforms.go#L25)):*
> ```Go
> //     a  b  c  x_off
> //     d  e  f  y_off
> //     g  h  i	z_off
> ```

nit, `z_off` is not column aligned


<!-- Sent from Reviewable.io -->

otan committed:
geomfn: consolidate affine transformations

* Introduce `applyOnGeomT`, which applies a function on a set
  of coordinates.
* Consolidate Translate, Scale and Rotate in terms of Affine.
  This is what PostGIS does, and removes a lot of duplicate code.
* Implement ST_Affine.
* Add unimplemented errors for remaining ST_Rotate items.

Release justification: low risk updates to to new functionality.

Release note (sql change): Implement ST_Affine.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53856)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola)

___
*[docs/generated/sql/functions.md, line 1239 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53856#-MGK5q6eADGsYFgSeVpf:-MGKSAXsAZD7JO3atW54:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c55485bc3ce493c80bf989d855bbf95b581aef3/docs/generated/sql/functions.md#L1239)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/apply/applied/
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
